Blog: keycloak (#122)
JankariTech/blog: src/assets/Keycloak/keycloak.md

---
title: Introduction to Keycloak
authorName: Amrita Shrestha
authorAvatar: https://avatars.githubusercontent.com/u/54478846?s=100&v=4
authorLink: https://github.com/amrita-shrestha
createdAt: March 26, 2024
tags: Keycloak
banner: https://blog.jankaritech.com/src/assets/keycloak/images/keycloak.png
---

## Keycloak
[Keycloak](https://github.com/keycloak/keycloak) is an open-source project for Red Hat SSO. It provides an `Identity and Access Management` solutions designed to secure application services.
Additionally, it enables users to authenticate through various identity providers and use fine-grained permissions for regulating access to Software as a Service (SaaS) applications.

> Keycloak alternatives are Auth0, Zluri, Microsoft Azure Active Directory, Okta or WSO2 Identity Server.

Reviewer: here you are talking currently about what keycloak can do, then suddenly about alternatives and again about what it can do. maybe shift the alternatives to the end of the chapter

It facilitates the creation of a user database with customizable roles and groups, offering functionalities such as user management, registration, and password policy enforcement.
This makes it a comprehensive Identity and Access Management (IAM) solution for contemporary applications. Keycloak's API enables the integration of all these features into your application
without the need for additional coding.

## Keycloak Features
![Keycloak features](/src/assets/Keycloak/images/keycloak-features.png)

1. `Identity Broker:`
   Identity Broker acts as a middleman, linking various service providers with different identity providers. It establishes trust with external identity providers to utilize their identities for accessing internal services offered by service providers.

>    Imagine a textile company, which has two Keycloak setups: one for customer and another for internal employees. If you want your employees to access external services without creating new accounts on customer Keycloak, you can use the customer Keycloak
>    as the mediator and the customer Keycloak as the source in this scenario.

Reviewer: ```suggestion
   Imagine a company, which has two Keycloak setups: one for customers and another for internal employees. If you want your employees to access external services without creating new accounts on the customer Keycloak, you can use the customer Keycloak as the mediator and the customer Keycloak as the source in this scenario.
```
I don't get it how the customer Keycloak is the mediator and the source.

2. `User Federation:`
   Keycloak provides the ability to integrate with existing LDAP or Active Directory servers. It also supports implementation of your own provider such as a relational database.

3. `Standard Protocols:`
   Keycloak is based on three standard protocols i.e., OpenID Connect, OAuth2.0, and SAML.

4. `Password Policies:`

>    Keycloak contain different password policies i.e., HashAlgorithm, Hashing Iterations, Digits, Lowercase Characters, Regular Expression and so on.

Reviewer 2: ```suggestion
   Keycloak contains different password policies i.e., HashAlgorithm, Hashing Iterations, Digits, Lowercase Characters, Regular Expression and so on.
```
The subject of this sentence (Keycloak) is in singular, therefore the verb needs an 's' :)

5. `Single-Sign On:`
   Keycloak facilitates a seamless login experience where users only need to sign in once to access multiple applications, eliminating the need for repeated logins.
   This not only streamlines user experience but also simplifies identity management for administrators.

6. `Social Login:`

>    Keycloak has built-in support for Google, GitHub, and Facebook which help us to use social identity providers.

Reviewer: ```suggestion
   Keycloak has built-in support to login via Google, GitHub, and Facebook, which helps us to use social identity providers.
```

## Keycloak Drawbacks
Despite its extensive features, Keycloak does have certain limitations. One of these is the need for a more varied implementation approach.
Additionally, the following are some of the drawbacks associated with Keycloak:

> 1. `Complex Server Deployment:` The manual process involved in deploying Keycloak on a server can be complicated, potentially affecting overall productivity.

Reviewer 2: I wouldn't use `xxx` syntax to highlight your points since this is for keywords/code snippets in plain text. I would rather make it bold or use the <strong> HTML tag for highlighting your points.

2. `Lack of Comprehensive Support Documentation:` Despite its robust functionalities, Keycloak could benefit from more efficient and extensive support documentation. This would make it easier for users to find suitable solutions to their problems.

> ## SetUp Keycloak using docker

Reviewer 3: ```suggestion
## Setup Keycloak using docker
```

Reviewer 2: Use same style for titles throughout your blog post (either all words capitalized or only first word and names)

1. Keycloak using docker image
```bash
docker run -p 8080:8080 -e KEYCLOAK_ADMIN=admin -e KEYCLOAK_ADMIN_PASSWORD=admin quay.io/keycloak/keycloak:24.0.2 start-dev
```

2. Keycloak with postgres database using docker-compose
   ![Setup Keycloak using docker compose](src/assets/keycloak/images/start-keycloak-with-docker-compose.png)

> 

Reviewer 2: Is that all that is used for setting up Keycloak? And are the first and the second point consecutive steps or alternative ways of doing this? If the later is the case, I find it confusing the way it is currently presented. Maybe an introduction sentence before the numbered list would help for clarification

> ## Why You Should Know Keycloak

Reviewer 3: ```suggestion
## Why You Should Know Keycloak ?
```

> ## Why You Should Know Keycloak

Reviewer 3: ```suggestion
## Why You Should Know Keycloak?
```

> 1. Open source

Reviewer 2: Since the other items in this list are sentences, try to also write a short sentence about this point.

> 2. Supports three different authentication protocols which give you the possibility to cover many applications with different security demands with a single tool.

Reviewer: ```suggestion
2. Supports three different authentication protocols which gives you the possibility to cover many applications with different security demands with a single tool.
```

> 2. Supports three different authentication protocols which give you the possibility to cover many applications with different security demands with a single tool.

Reviewer 2: ```suggestion
2. Keycloak supports three different authentication protocols which give you the possibility to cover many applications with different security demands with a single tool.
```
or 

```suggestion
2. It supports three different authentication protocols which give you the possibility to cover many applications with different security demands with a single tool.
```
subject in this sentence is missing

> 3. Provides web-based GUI which makes any configurations changes easier.

Reviewer: easier than what?
Or do you mean to say "easy"?

> 4. Huge community support

Reviewer 2: As a reader (and potential future Keycloak user) it would be nice to have some links to documentation or forums where the community mingles :)

> ## When It May Not Be the Best Choice

Reviewer 3: ```suggestion
## When It May Not Be the Best Choice?
```

> ## When It May Not Be the Best Choice

Reviewer 4: ```suggestion
## Why it May Not Be the Best Choice?
```

Reviewer 2: Depends on what you want to say... When you use 'why' it means that it is not the best choice. By using 'when' it means that in certain situations it is the best choice and in other situations it's maybe not a good choice...
